src/c/modularizace/linker.md

```diff
@@ -0,0 +1,97 @@
+Programy v *C* se skládají z jedné či více tzv. **jednotek překladu** (*translation unit*). Jedná se
+o nezávislé komponenty, ze kterých je nakonec vytvořen cílový program. Každá jednotka je obvykle
+tvořena jedním zdrojovým souborem (obvykle s příponou `.c`). Pří překladu **překladač** převede
+jednotku ze zdrojového kódu v *C* do instrukcí procesoru, tzv. **objektového kódu** (*object code*).
+
+Pokud chceme překladačem `gcc` (pouze) přeložit zdrojový soubor do objektového kódu (resp.
+objektového souboru), použijeme přepínač `-c`:
+```bash
+$ gcc -c soubor.c
+```
+Pokud nezadáme název výstupu pomocí přepínače `-o`, tak `gcc` implicitně vytvoří objektový soubor
+`<nazev-vstupu>.o` (tj. zde `soubor.o`).
+
+Jednotky překladu jsou na sobě nezávislé, můžeme tedy každou jednotku (zdrojový soubor) přeložit
+zvlášť:
+```bash
+$ gcc -c a.c
+$ gcc -c b.c
+...
+```
+
+Jak ale nyní jednotlivé soubory propojíme? Aby vůbec mělo rozdělení do více jednotek (souborů) smysl,
+tak musíme mít možnost v jednom souboru používat kód (např. funkce nebo globální proměnné), který je
+nadefinovaný v jiném souboru. V *C* toto propojení jednotek neprobíhá při překladu, ale až v následné
+fázi nazývané linkování.
+
+## Linkování programu
+Jakmile přeložíme všechny naše zdrojové soubory postupně do objektových souborů, potřebujeme z nich
+vytvořit finální spustitelný soubor, což je práce programu nazývaného **linker**. Linker obdrží
+seznam všech (již přeložených) objektových souborů, ze kterých se má program skládat, propojí je
+dohromady a vytvoří z nich spustitelný soubor.
+
+Jak propojení jednotlivých souborů probíhá? Představme si například, že v souboru `a.c` voláme
+funkci `foo`, která v tomto souboru neexistuje. Při překladu tohoto souboru překladač vytvoří
+objektový soubor `a.o`, ve kterém bude uložena informace, že voláme funkci `foo`. Dejme tomu, že
+tato funkce existuje v souboru `b.c`, který je přeložen do objektového souboru `b.o`. Při linkování
+linker obdrží seznam všech obj. souborů, tedy `a.o` i `b.o`. Když narazí na informaci, že z `a.o`
+chceme volat funkci `foo`, pokusí se tuto funkci naleznout v některém z předaných objektových
+souborů:
+- Pokud jej nenalezne, tak vypíše chybu a program se nepřeloží[^1].
+- Pokud jej nalezne (v tomto případě v `b.o`), tak volání funkce "propojí" tak, aby se volala správná
+funkce původně vytvořená v `b.c`.
+
+[^1]: V takovém případě byste se setkali s chybou `undefined reference to 'foo'`.
+
+Manuální použití linkeru[^2] je relativně složité, proto i linker budeme používat přes `gcc`. Tomu
+můžeme předat sadu objektových souborů a on se postará o správné zavolání linkeru, který je spojí
+a vytvoří finální spustitelný soubor:
+```bash
+$ gcc a.o b.o -oprogram
+```
+
+[^2]: Na Linuxu lze naleznout například linker `ld`.
+
+Při finálním linkování programu také dochází ke kontrole toho, jestli je v některém z objektových
+souborů obsažena funkce `main`, aby program věděl, kde má začít své vykonávání.
+
```

Comment:
No, mohlo by se pro uplnost ukazat i `ldd` a `nm`, ale... asi by to byla opet zabijacka. Na prednasce jsem to loni ukazoval, ale jestli z toho neco bylo, tak to nevim.

Reply:
`ldd` tam je, počej si :) `nm`.. asi bych implementační detaily nechal na jindy.